<a href="https://colab.research.google.com/github/me2527/WhiteHatJrLottery/blob/main/2020_09_12_AditiDwivediCapstoneProject13_WhitehatJrLottery.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 13: WhiteHat Jr Lottery

---

### Context

Many people have the fantasy of winning a lottery. The lottery seems a speedy way of becoming rich. Lotteries refer to games where we need to guess which numbers will be drawn from a pool of options. They are designed in such a way that there are extremely low chances of winning a jackpot amount.

During advertising, many companies that run lotteries target individuals belonging to the low income and debt-ridden groups. Lotteries can incur heavy losses to your income yet studies revealed that financially challenged people are more likely to purchase lottery tickets as compared to the wealthy people. They spend a large amount of their meagre income to purchase lottery tickets. These lotteries not only exploit the desires of the underprivileged of becoming rich but also worsens their current financial condition. Lottery houses use the lottery mathematics which is nothing but the probability and expected value to design their lottery games in a way that chances of winning the lottery are extremely low.

Let us try to understand the underlying math involved in a Lottery game.

---

### Problem Statement

Imagine a situation in which WhiteHat Jr announced its lottery game. The amount of the jackpot prize to be won is INR 25,000,000 (or INR 25 million).

Suppose you want to play this lottery and you bought a ticket for it which costs INR 250. It has two identical leaves or copies; one is for yourself and another for WhiteHat Jr.

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lottery_ticket.png' width=500>

You have to fill both the copies and give one back to WhiteHat Jr. Each leaf has 70 boxes numbered from 1 to 70. You have to mark 10 different numbers out of those 70 numbers in both the leaves and submit the second copy of the filled ticket back to WhiteHat Jr.

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lottery_ticket_filled.png' width=500>

After a week in a lucky draw, 10 balls will be drawn randomly one-by-one from an urn of 70 balls. Each ball is labelled from 1 to 70. Once a ball is drawn, it doesn't go back inside the urn.

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lottery_machine.jpg' width=500>

If the number printed on the balls drawn from the urn matches with the numbers that you marked on your lottery ticket, then you will win a prize but there are some conditions involved. Here's the list of all the conditions of winning the prizes:

- If all 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win **INR 25,000,000, the jackpot prize**.

- If 9 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win INR 2,500,000.

- If 8 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win INR 250,000.

- If 7 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win INR 25,000.

- If 6 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win INR 2,500.

- If 5 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the balls drawn, you will win INR 250.

- In all other cases, you win nothing.

Find out the amount you are expected to win or lose on an average per lottery ticket.

**Note:** The student must first attempt the problem on their own and then they should take the help of the teacher to solve it.

---

#### Getting Started

Follow the steps described below to solve the project:

1. Click on the link provided below to open the Colab file for this project.
   
   https://colab.research.google.com/drive/1C5I6usaFMdWtS6Y5ts1KBABv7fZUn-Ya?usp=sharing

2. Create the duplicate copy of the Colab file. Here are the steps to create the duplicate copy:

    - Click on the **File** menu. A new drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/0_file_menu.png' width=500>

    - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/1_create_colab_duplicate_copy.png' width=500>

     - After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_CapstoneProject13** format. 

3. Now, write your code in the prescribed code cells.

---

### Recommended Approach

1. List down the possible cases in which the numbers marked on the lottery ticket matches with the numbers on the 10 balls drawn from the urn. Eg.

   - **Possibility 1:** 0 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the 10 balls drawn. ${}^{10}C_0 = \frac{10!}{0! \times 10!} = 1$ way as shown in the image below.

     <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/no_numbers_match.png' width=400>

     **Note:** The **green** colour denotes that the numbers marked on the ticket matches with the ones on the balls drawn. The **red** colour denotes that the numbers marked on the ticket **do not** match with the ones on the balls drawn. 

   - **Possibility 2:** 1 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the 10 balls drawn. This can be done in ${}^{10}C_1 = \frac{10!}{1! \times 9!} = \frac{10 \times 9!}{9!} = 10$ ways as shown in the image below. 

     <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/one_numbers_match.png' width=400>

    

   - **Possibility 3:** 2 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the 10 balls drawn. This can be done in ${}^{10}C_2 = \frac{10!}{2! \times 8!} = \frac{10\times 9 \times 8!}{2! \times 8!} = 45$ ways as shown in the image below.

     <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/two_numbers_match.png' width=400>

    .

    .

    .

   - **Possibility 11:** 10 out of the 10 numbers that you marked on your lottery ticket matches with the ones on the 10 balls drawn. This can be done in ${}^{10}C_{10} = \frac{10!}{10! \times 0!} = 1$ way as shown in the image below.

     <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/all_numbers_match.png' width=400>

2. Find out the probabilities of each possibility.

3. Multiply the amount to be won or lost with their corresponding probabilities. 

4. Add all the products (obtained in step 3) to get the expected amount to be won or lost after buying a lottery ticket.

5. Implement the solution using Python.

---

In [ ]:
# Create a list containing the number of favourable outcomes per draw for random variables X = 0 to X = 10.
favourable_outcomes =[]

for i in range(0,11):
  f=[]
  for j in range(0,10):
    if j<i:
      f.append(1)
    else:
      f.append(69-j)
  favourable_outcomes.append(f)
favourable_outcomes


[[69, 68, 67, 66, 65, 64, 63, 62, 61, 60],
 [1, 68, 67, 66, 65, 64, 63, 62, 61, 60],
 [1, 1, 67, 66, 65, 64, 63, 62, 61, 60],
 [1, 1, 1, 66, 65, 64, 63, 62, 61, 60],
 [1, 1, 1, 1, 65, 64, 63, 62, 61, 60],
 [1, 1, 1, 1, 1, 64, 63, 62, 61, 60],
 [1, 1, 1, 1, 1, 1, 63, 62, 61, 60],
 [1, 1, 1, 1, 1, 1, 1, 62, 61, 60],
 [1, 1, 1, 1, 1, 1, 1, 1, 61, 60],
 [1, 1, 1, 1, 1, 1, 1, 1, 1, 60],
 [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

In [ ]:
# Create a list containing probabilities of the random variable X = 0 to X = 10.
prob_favourable_outcomes = []
for i in range(0,11):
  p=[]
  for j in range(0,10):
    if j<i:
      p.append(1 / (70 - j))
    else:
      p.append((69 - j) / (70 - j))
  prob_favourable_outcomes.append(p)
prob_favourable_outcomes

[[0.9857142857142858,
  0.9855072463768116,
  0.9852941176470589,
  0.9850746268656716,
  0.9848484848484849,
  0.9846153846153847,
  0.984375,
  0.9841269841269841,
  0.9838709677419355,
  0.9836065573770492],
 [0.014285714285714285,
  0.9855072463768116,
  0.9852941176470589,
  0.9850746268656716,
  0.9848484848484849,
  0.9846153846153847,
  0.984375,
  0.9841269841269841,
  0.9838709677419355,
  0.9836065573770492],
 [0.014285714285714285,
  0.014492753623188406,
  0.9852941176470589,
  0.9850746268656716,
  0.9848484848484849,
  0.9846153846153847,
  0.984375,
  0.9841269841269841,
  0.9838709677419355,
  0.9836065573770492],
 [0.014285714285714285,
  0.014492753623188406,
  0.014705882352941176,
  0.9850746268656716,
  0.9848484848484849,
  0.9846153846153847,
  0.984375,
  0.9841269841269841,
  0.9838709677419355,
  0.9836065573770492],
 [0.014285714285714285,
  0.014492753623188406,
  0.014705882352941176,
  0.014925373134328358,
  0.9848484848484849,
  0.9846153846153847,
  0.

In [ ]:
# Function to calculate the factorial value of a number 'n'.
def factorial(num):
  factorial = 1
  if num < 0:
    return "Undefined."
  else:
    while num > 0:
      factorial *= num
      num -= 1
    return factorial

In [ ]:
# Calculate the amount you are expected to will win or lose after buying the WhiteHat Jr lottery ticket.
ticket_price = 250
prize = 250 # Using a loop, the prize value will go from 250 to 25,000,000 by a factor of 10.
expected_value = 0

for i in range(5):
  prob_random_variable = 1 # Calculate the probability of a random variable X = 0 to X = 4.
  for j in range(len(prob_favourable_outcomes[i])):
    prob_random_variable *= prob_favourable_outcomes[i][j]
  # Multiply probability with the number of ways the numbers on the ticket can match with the numbers on the balls drawn.
  prob_random_variable *= factorial(10) / (factorial(i) * factorial(10 - i))
  print(f"\nProbability of {i} matches = {prob_random_variable}")
  expected_value += -ticket_price * prob_random_variable
  print(f"Expected value after {i} matches = {expected_value}")

print("\nEnd of first 5 iterations.\n")

for i in range(5, 11):
  prob_random_variable = 1 # Calculate the probability of a random variable X = 5 to X = 10.
  for j in range(len(prob_favourable_outcomes[i])):
    prob_random_variable *= prob_favourable_outcomes[i][j]
  # Multiply probability with the number of ways the numbers on the ticket can match with the numbers on the balls drawn.
  prob_random_variable *= factorial(10) / (factorial(i) * factorial(10 - i))
  print(f"\nProbability of {i} matches = {prob_random_variable}")
  expected_value += prize * prob_random_variable
  print(f"For {i} matches, prize = INR {prize:,.2f}") # Print the new prize value such that the digits are separated by comma.
  print(f"Expected value after {i} matches = {expected_value}")
  prize *= 10
 


Probability of 0 matches = 0.8571428571428573
Expected value after 0 matches = -214.28571428571433

Probability of 1 matches = 0.12422360248447203
Expected value after 1 matches = -245.34161490683235

Probability of 2 matches = 0.008220679576178298
Expected value after 2 matches = -247.39678480087693

Probability of 3 matches = 0.0003271912269125691
Expected value after 3 matches = -247.47858260760506

Probability of 4 matches = 8.675524956015087e-06
Expected value after 4 matches = -247.48075148884408

End of first 5 iterations.


Probability of 5 matches = 1.601635376495093e-07
For 5 matches, prize = INR 250.00
Expected value after 5 matches = -247.48071144795966

Probability of 6 matches = 2.085462729811319e-09
For 6 matches, prize = INR 2,500.00
Expected value after 6 matches = -247.48070623430283

Probability of 7 matches = 1.891576172164462e-11
For 7 matches, prize = INR 25,000.00
Expected value after 7 matches = -247.48070576140879

Probability of 8 matches = 1.1440984912285052

**Q:** What is the expected amount to be won or lost after buying a lottery ticket?

**A:** -247.48

---

### Submitting the Project

Follow the steps described below to submit the project.

1. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/2_share_button.png' width=500>

2. In the dialog box, click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/3_copy_link.png' width=500>


3. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_CapstoneProject13**) of the notebook will get copied 

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/4_copy_link_confirmation.png' width=500>

4. Go to your dashboard and click on the **My Projects** option.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/5_student_dashboard.png' width=800>

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/6_my_projects.png' width=800>

5. Click on the **View Project** button for the project you want to submit.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/7_view_project.png' width=800>

6. Click on the **Submit Project Here** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/8_submit_project.png' width=800>

7. Paste the link to the project file named as **YYYY-MM-DD_StudentName_CapstoneProject13** in the URL box and then click on the **Submit** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/9_enter_project_url.png' width=800>


---